cli/internal/plugin/builtin/oci/resource_repository.go

```diff
@@ -0,0 +1,157 @@
+	t := resource.Access.GetType()
+	obj, err := p.scheme.NewObject(t)
+	if err != nil {
+		return nil, fmt.Errorf("error creating new object for type %s: %w", t, err)
+	}
+	if err := p.scheme.Convert(resource.Access, obj); err != nil {
+		return nil, fmt.Errorf("error converting access to object of type %s: %w", t, err)
+	}
+	switch access := obj.(type) {
+	case *v1.OCIImage:
+		baseURL, err := ociImageAccessToBaseURL(access)
+		if err != nil {
+			return nil, fmt.Errorf("error creating oci image access: %w", err)
+		}
+
+		repo, err := p.getRepository(&ociv1.Repository{
+			BaseUrl: baseURL,
+		}, credentials)
+		if err != nil {
+			return nil, fmt.Errorf("error creating repository: %w", err)
+		}
+
+		resource = resource.DeepCopy()
+		resource.Access = access
+
+		resource, err := repo.ProcessResourceDigest(ctx, resource)
+		if err != nil {
+			return nil, fmt.Errorf("error downloading resource: %w", err)
+		}
+
+		return resource, nil
+	default:
+		return nil, fmt.Errorf("unsupported type %s for downloading the resource", t)
+	}
+}
+
+func (p *ResourceRepositoryPlugin) getIdentity(obj runtime.Typed) (runtime.Identity, error) {
+	switch access := obj.(type) {
+	case *v1.OCIImage:
+		baseURL, err := ociImageAccessToBaseURL(access)
+		if err != nil {
+			return nil, fmt.Errorf("error creating oci image access: %w", err)
+		}
+		identity, err := runtime.ParseURLToIdentity(baseURL)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing URL to identity: %w", err)
+		}
+		identity.SetType(runtime.NewVersionedType(ociv1.Type, ociv1.Version))
+		return identity, nil
+	default:
+		return nil, fmt.Errorf("unsupported type %s for getting identity", obj.GetType())
+	}
+}
+
+func (p *ResourceRepositoryPlugin) DownloadResource(ctx context.Context, resource *descriptor.Resource, credentials map[string]string) (blob.ReadOnlyBlob, error) {
+	t := resource.Access.GetType()
+	obj, err := p.scheme.NewObject(t)
+	if err != nil {
+		return nil, fmt.Errorf("error creating new object for type %s: %w", t, err)
+	}
+	if err := p.scheme.Convert(resource.Access, obj); err != nil {
+		return nil, fmt.Errorf("error converting access to object of type %s: %w", t, err)
+	}
+	switch access := obj.(type) {
+	case *v1.OCIImage:
+		baseURL, err := ociImageAccessToBaseURL(access)
+		if err != nil {
+			return nil, fmt.Errorf("error creating oci image access: %w", err)
+		}
+
+		repo, err := p.getRepository(&ociv1.Repository{
+			BaseUrl: baseURL,
+		}, credentials)
+		if err != nil {
+			return nil, fmt.Errorf("error creating repository: %w", err)
+		}
+
+		b, err := repo.DownloadResource(ctx, resource)
+		if err != nil {
+			return nil, fmt.Errorf("error downloading resource: %w", err)
+		}
+
+		return b, nil
+	default:
+		return nil, fmt.Errorf("unsupported type %s for downloading the resource", t)
+	}
+}
+
+// TODO Repeated calls with separate credentials will always use the first credentials set.
```

Comment:
Is still still true? 🤔

Reply:
Will have to double check this. I think this might still be the case. We dont dynamically override credentials I believe